rust-runtime/aws-smithy-http/src/byte_stream/bytestream_util.rs

```diff
@@ -8,35 +8,169 @@ use futures_core::{ready, Stream};
+///     let bytestream = ByteStream::read_from()
+///         .path("docs/some-large-file.csv")
+///         // Specify the size of the buffer used to read the file (in bytes, default is 4096)
+///         .buffer_size(32_784)
+///         // Specify the length of the file used (skips an additional call to retrieve the size)
+///         .file_size(123_456)
+///         .build()
+///         .await
+///         .expect("valid path");
+///     GetObjectInput { body: bytestream }
+/// }
+/// # }
+/// ```
+pub struct FsBuilder {
+    file: Option<tokio::fs::File>,
+    path: Option<PathBuf>,
+    file_size: Option<u64>,
+    buffer_size: usize,
+}
+
+impl FsBuilder {
+    /// Create a new [`FsBuilder`] (using a default read buffer of 4096 bytes).
+    ///
+    /// You must then call either [`file`](FsBuilder::file) or [`path`](FsBuilder::path) to specify what to read from.
+    pub fn new() -> Self {
+        FsBuilder {
+            file: None,
+            path: None,
+            file_size: None,
+            buffer_size: DEFAULT_BUFFER_SIZE,
+        }
+    }
+
+    /// Sets the path to read from.
+    ///
+    /// NOTE: The resulting ByteStream (after calling [build](FsBuilder::build)) will be retryable.
+    /// The returned ByteStream will provide a size hint when used as an HTTP body.
+    /// If the request fails, the read will begin again by reloading the file handle.
+    pub fn path(mut self, path: impl AsRef<std::path::Path>) -> Self {
+        self.path = Some(path.as_ref().to_path_buf());
+        self
+    }
+
+    /// Sets the file to read from.
+    ///
+    /// NOTE: The resulting ByteStream (after calling [build](FsBuilder::build)) will not be a retryable ByteStream.
+    /// For a ByteStream that can be retried in the case of upstream failures, use [`FsBuilder::path`](FsBuilder::path).
+    pub fn file(mut self, file: tokio::fs::File) -> Self {
+        self.file = Some(file);
+        self
+    }
+
+    /// Specify the length of the file to read (in bytes).
+    ///
+    /// By pre-specifying the length of the file, this API skips an additional call to retrieve the size from file-system metadata.
+    pub fn file_size(mut self, file_size: u64) -> Self {
+        self.file_size = Some(file_size);
+        self
+    }
+
+    /// Specify the size of the buffer used to read the file (in bytes).
+    ///
+    /// Increasing the read buffer capacity to higher values than the default (4096 bytes) can result in a large reduction
+    /// in CPU usage, at the cost of memory increase.
+    pub fn buffer_size(mut self, buffer_size: usize) -> Self {
+        self.buffer_size = buffer_size;
+        self
+    }
+
+    /// Returns a [`ByteStream`](crate::byte_stream::ByteStream) from this builder.
+    /// NOTE: If both [`file`](FsBuilder::file) and [`path`](FsBuilder::path) have been called for this FsBuilder, `build` will
+    /// behave read from the path specified by [`path`](FsBuilder::path).
```

Comment:
Would you mind leaving a short note here about how this can panic if no setters were called before building? Something like
```suggestion
    /// Returns a [`ByteStream`](crate::byte_stream::ByteStream) from this builder.
    /// NOTE: If both [`file`](FsBuilder::file) and [`path`](FsBuilder::path) have been called for this FsBuilder, `build` will
    /// behave read from the path specified by [`path`](FsBuilder::path).
    ///
    /// # Panics
    ///
    /// Panics if neither of the `file` or`path` setters were called.
```